extensions/2.0/Khronos/KHR_materials_volume/README.md

```diff
@@ -0,0 +1,172 @@
+
+The light rays are reflected or refracted at the micro-facets on the surface, taking the roughness of the material into account. The index of refraction used for computing the refraction direction is taken from the BSDF, potentially set by `KHR_materials_ior`.
+
+<figure style="text-align:center">
+<img src="./figures/volume.svg"/>
+<figcaption>Object with homogeneous volume and an index of refraction of 1.5.</figcaption>
+</figure>
+
+## Extending Materials
+
+The volumetric properties are defined by adding the `KHR_materials_volume` extension to any glTF material.
+
+```json
+materials: [
+    {
+        "extensions": {
+            "KHR_materials_volume": {
+                "thinWalled": false,
+                "attenuationCoefficient": [ 78.03, 17.83, 126.29 ],
+                "volumeAlbedo": [ 0.29, 0.82, 0.99 ],
+                "subsurfaceAnisotropy": 0.0
+            }
+        }
+    }
+]
+```
+
+Note that there are no textures in this extension, the medium is homogeneous.
+
+## Properties
+
+The extension provides a Boolean parameter to switch between thin-walled and volumetric behavior. In addition, it defines parameters to describe the medium and the phase function inside the volume.
+
+| | Type | Description | Required |
+|-|------|-------------|----------|
+| **thinWalled** | `Boolean` | Material describes either a two-sided surface enclosing an infinitely thin volume (true, the default) or a one-sided surface with a volume below it, defined by the mesh geometry (false). | No, default: `true` |
+| **attenuationCoefficient** | `number[3]` | Attenuation coefficient (also called extinction coefficient) in inverse meters. | No, default: `[0.0, 0.0, 0.0]` |
+| **volumeAlbedo** | `number[3]` | Single-scattering albedo inside the volume. | No, default: `[0.0, 0.0, 0.0]` |
+| **subsurfaceAnisotropy** | `number` | Anisotropy of the phase function in range [-1, 1] | No, default: `0` |
+| **thicknessTexture** | `textureInfo` | A texture that defines the thickness of the volume beneath the surface. The values are sampled from the G channel. May be used by renderers that are not able to derive the thickness from the actual geometry, otherwise it should be ignored. | No |
+| **minThickness** | `float` | When using a thickness texture, a texture value of 0.0 is interpreted to be `minThickness`. When a constant thickness is used, the thickness will be `minThickness`. | No, default: `0` |
+| **maxThickness** | `float` | When using a thickness texture, a texture value of 1.0 is interpreted to be `maxThickness`. Values between 0.0 and 1.0 are linearly scaled between minThickness and `maxThickness`. | No, default: `1` |
```

Comment:
Elsewhere, glTF materials allow a texture or a factor for inputs. I realize we may not want textures for `attenuationCoefficient`, `volumeAlbedo`, and `subsurfaceAnisotropy` for complexity reasons. But is there a reason to exclude a factor for `thickness`? I think you could parameterize it in the usual way, and remove the need for `thinWalled` boolean at the same time:

- `thicknessFactor`: Thickness of volume beneath the surface, 0-1, used to lerp between `minThickness` and `maxThickness`. Default is 0.
- `thicknessTexture`: Thickness of volume beneath the surface, 0-1, used to lerp between `minThickness` and `maxThickness`. Linear multiplier against `thicknessFactor`.
- `minThickness`: No change.
- `maxThickness`: No change.

Note that because the default is 0, the transmission is "off" by default (which we want) without the need for a separate property. If we did keep the separate property for that purpose, I think I somewhat prefer the terminology that Filament uses: `refractionType: 'thin' | 'volume'`.

Also — should `minThickness` and `maxThickness` specify units, e.g. meters?

Reply:
I like this.
I think you mean the **volume** is off my default. Transmission shouldn't be affected.
Also, +1 for metres as the default unit for thickness. That matches the default unit of the rest of the glTF file.

Reply:
That would mean the thickness is computed as follows, right?

```
thickness = lerp(minThickness, maxThickness, thicknessFactor * thicknessTexture)
```

I like it, it's much more similar to the usual texture parameters used everywhere else.

However I didn't get why thicknessFactor=0 would disable the volume. Wouldn't it mean that a volume is still present, just with constant thickness = minThickness?

Maybe `surfaceType: 'thin' | 'volumeBoundary'` instead of `refractionType`? Probably also not optimal, but it would stress the fact that not only refraction is affected by the type, but also attenuation and subsurface scattering. Nitpicking, I know, but having to choose those names was the reason for me switching to a Boolean instead of an enum :-)

I will add information about the unit, it should definitely be meters for min and max thickness.

Reply:
> I think you mean the volume is off my default.

Oops, yes.

> Wouldn't it mean that a volume is still present, just with constant thickness = minThickness?

Discussed offline – the default for both `thicknessFactor` and `minThickness` will be zero, so the volume is disabled by default. But you're correct that increasing either of these values will cause volume to be enabled.

Reply:
Another suggestion from @UX3D-nopper that sounds useful — we could omit `minThickness` and `maxThickness`, having only two parameters here:

- `thicknessFactor`
- `thicknessTexture`

The product of the two would be defined in units of meters. When `thicknessFactor=0` (default), volume is disabled. Note that this simpler parameterization loses precision _if_ the minimum thickness of the volume is >> 0. Is it likely for minimum thickness to be non-zero in real use cases? If we need that case, the addition of a `thicknessOffset` parameter would restore the precision while keeping a simpler parameterization. In that case the thickness would be computed as:

```
thickness = thicknessFactor * thicknessTexture + thicknessOffset
```

Reply:
There are several tools that already support loading or baking thickness maps:

- [Substance Painter](https://docs.substance3d.com/spdoc/thickness-142213479.html) (both)
- [Marmoset Toolbag](https://marmoset.co/posts/toolbag-baking-tutorial/) (bake)
- [Sketchfab](https://help.sketchfab.com/hc/en-us/articles/204429595-Materials-PBR-#sss) (load)
- [Blender](https://docs.blender.org/manual/en/latest/render/shader_nodes/input/ao.html) (bake, via AO "inside")

I could only find the units for Substance Painter, which has two settings:

1. units used in "Max Frontal Distance" et "Max Rear Distance" are the ones defined when exporting your mesh (m, cm or whatever you did set up in your scene export)
2. `units are relative to the bounding box of the object, 1.0 being the diagonal length of the bounding box`.

If there is some consistency among these tools, that might take precedence over some of the suggestions above, for me.

Reply:
That's very good information, thanks for sharing! I had a look at [refraction in Babylon.js](https://doc.babylonjs.com/how_to/physically_based_rendering_master) and [Filament](https://google.github.io/filament/Materials.html#materialmodels/litmodel/absorption) to see which of the two modes they consume. Documentation is a bit sparse, but from what I understand they both take the thickness in meters. Babylon.js uses factor and offset as suggested in your previous comment: `minimumThickness + thicknessTexture.r * maximumThickness`

Reply:
Have we resolved this thread? I don't know whether something like `minThickness` and `maxThickness` (or `thicknessOffset`) are really important, but just want to verify that we're OK with omitting them as this extension moves to completion.

Reply:
The `thicknessFactor` is now specified in meters, although I started a separate thread comparing it to `attenuationDistance`.  I think this old thread can be resolved, as there is no more min/max thickness.

Reply:
Also I think we ruled out "min" thinkness, on the thought that any sharp edge, such as along the edge of a cube, the thickness approaches zero the closer you get to the edge.  So "min" didn't seem useful in that context, but maybe the Babylon folks have different thoughts about that.

For "max", it was just included in the thicknessFactor by allowing values higher than 1.0.

Reply:
> The thicknessFactor is now specified in meters,

I can't find this in the specification, is this still current? The language in the extension specification seems unclear to me, and could be interpreted several ways:

>  The value is relative to the local coordinate system of the mesh's node.